Test for outdated spec URLs that aren't in w3c/browser-specs

Elchi3 opened the pull request:
We need to keep our specification URLs current to always provide the recommended and most up-to-date spec_urls as specified by w3c/browser-specs.

Therefore we should check all of our spec_urls against the w3c/browser-spec repository. The repo includes almost all of the BCD spec hosts but a few are missing still and so I've added a temporal allow list.

This PR fails because in main BCD doesn't use the recommend spec for ECMAScript right now. https://github.com/mdn/browser-compat-data/pull/10615 fixes that.

This adds a dev dependency to w3c/browser-specs. Sometimes, when dependabot asks us to update to a new version of browser-specs, it might fail the build because there a recommended spec from browser-specs has changed. Just like it is the case now with ECMAScript. So the idea is that this will add a mechanism that keeps spec urls up to date and relevant.

(I've never written a mocha test for BCD, please take it apart. Edit: I'm not happy with the error message, so maybe we can improve that if the overall approach of this PR makes sense.)

Elchi3 commented:
> This PR fails because in main BCD doesn't use the recommend spec for ECMAScript right now. #10615 fixes that.

I merged that PR and so this PR passes now. We now only use specs listed by w3c/browser-specs (minus the temporal allow list introduced in this PR).

ddbeck commented on test/spec-urls.test.js:
```diff
@@ -0,0 +1,96 @@
+
+describe('spec_url data', () => {
+  it('spec_url specifications only use allow listed hosts by w3c/browser-specs (and those we allow list for now)', () => {
+    const toVisit = [
+      'api',
+      'css',
+      'html',
+      'http',
+      'javascript',
+      'mathml',
+      'svg',
+      'webdriver',
+    ];
+
+    let specURLs = [];
+    for (const key of toVisit) {
+      visit(
+        (path, feature) => {
+          if (feature.spec_url) {
+            if (Array.isArray(feature.spec_url)) {
+              specURLs.push(...feature.spec_url);
+            } else {
+              specURLs.push(feature.spec_url);
+            }
+          }
+        },
+        {
+          entryPoint: key,
+        },
+      );
+    }
```
Since we're not using any of the control flow that `visit()` provides, I'd probably use `walk()` instead. Something like this?

```js
const specURLs = [];
for (const { compat } of walk()) {
  const { spec_url } = compat;
  const specs = [].concat(spec_url || []); // coerce spec_url to array, or empty array if undefined
  specURLs.push(...specs);
}
```

```suggestion
  const specURLs = [];
  for (const { compat } of walk()) {
    const { spec_url } = compat;
    const specs = [].concat(spec_url || []); // coerce spec_url to array, or empty array if undefined
    specURLs.push(...specs);
  }
```

ddbeck commented on test/spec-urls.test.js:
```diff
@@ -0,0 +1,96 @@
+      );
+    }
+
+    let specsNotInBrowserSpecs = [];
```
Starting from here, I found the approach to be kinda indirect. I think it'd be a lot more obvious what you're doing if you constructed a single allow list, then checked each URL. Something like this…

```
const specsFromBrowserSpecs = specData.map(
  spec => spec.url || spec.nightly.url || spec.series.nightlyUrl,
);
const specsExceptions = [
// .. long list here ..
]
const allowList = new Set([...specsFromBrowserSpecs, ...specsExceptions]);

const rejectedSpecs = [];
for (const spec of specURLs) {
  // some stuff here that checks whether the spec (minus the hash) is in the set of allowed specs
}
```

ddbeck reviewed (changes requested):
Thank you, Florian! A couple of ideas for you in line comments.

More broadly, it's probably not optimal for us to introduce a lot of standalone linting like this, lest we traverse the whole tree a dozen times, which is rather slow. But that shouldn't stop us now. That said, the next time we want to add something like this, we should give some thought to walking the tree once and, at each feature, dispatching to individual checks.

Elchi3 commented on test/spec-urls.test.js:
```diff
@@ -0,0 +1,74 @@
+
+    const specsExceptions = [
+      'https://wicg.github.io/controls-list/',
+      'https://w3c.github.io/webrtc-extensions/',
```
I can remove this exception if https://github.com/mdn/browser-compat-data/pull/10725 is merged

Elchi3 replied:
Was merged, removed this.

Elchi3 commented on test/spec-urls.test.js:
```diff
@@ -0,0 +1,74 @@
+      'https://wicg.github.io/controls-list/',
+      'https://w3c.github.io/webrtc-extensions/',
+      'https://w3c.github.io/setImmediate/',
+      'https://datatracker.ietf.org/doc/html/rfc2397',
```
Most of these will no longer be exceptions if https://github.com/w3c/browser-specs/issues/280 is fixed.

Elchi3 commented:
> Thank you, Florian! A couple of ideas for you in line comments.

Thanks Daniel! I think I addressed your comments. Thanks for the great guidance here :)

> More broadly, it's probably not optimal for us to introduce a lot of standalone linting like this, lest we traverse the whole tree a dozen times, which is rather slow. But that shouldn't stop us now. That said, the next time we want to add something like this, we should give some thought to walking the tree once and, at each feature, dispatching to individual checks.

Yeah, I totally agree. I think it is a bit unfortunate that the test outputs all of the files and even marks them as passing and then at the end another walk takes places and might tell you "oops, nope, I found more and this is actually wrong". So, yes, a general purpose traverse would be cool and then some of the current lints could maybe find their way into mocha based tests, too.

Elchi3 commented on test/spec-urls.test.js:
```diff
@@ -0,0 +1,73 @@
+      'https://webassembly.github.io/threads/js-api/',
+      'https://tc39.es/proposal-hashbang/out.html',
+      'https://mathml-refresh.github.io/mathml/',
+      'https://www.w3.org/TR/xpath-31/',
+      'https://www.w3.org/TR/xslt-30/',
```
These two can be removed once we removed XPath and XSLT in https://github.com/mdn/browser-compat-data/pull/9830

ddbeck replied:
Upon merging, I'll update #9830 to drop these two.

Elchi3 commented on test/spec-urls.test.js:
```diff
@@ -0,0 +1,73 @@
+
+    const specsExceptions = [
+      'https://wicg.github.io/controls-list/',
+      'https://w3c.github.io/setImmediate/',
```
Should be removed if/when https://github.com/mdn/browser-compat-data/pull/10746 lands

ddbeck reviewed (approved):
Looks good. Thank you! 🎉
